[core]: Adds support for on_error to abort further event handlers on exceptions

prologic opened the pull request:
Closes #128

prologic committed:
[core]: Add support for on_error to abort further event handlers on exceptions

Fixes #128

spaceone commented on circuits/core/events.py:
```diff
@@ -18,6 +18,10 @@ class Event(object):
     waitingHandlers = 0
 
+    # Behaviour for handling errors:
+    # One of "abort", "igore" (default)
```
typo: "igore"

apollo13 replied:
If "ignore" is the default, wouldn't it make sense to set `on_error` to that instead of None?

spaceone reviewed (changes requested):
I need to have this flag also for the specific handler, is this possible?

apollo13 commented on circuits/core/manager.py:
```diff
@@ -706,6 +706,14 @@ def _dispatcher(self, event, channels, remaining):  # noqa
+                elif event.on_error == "ignore":
+                    continue
+                else:
+                    continue
```
Since strings are hard to handle (prone to typos), I think it would be better to check if `on_error` is in the allowed options and raise otherwise. What about using constants for those strings?

spaceone replied:
I am unsure about using constants because everybody needs to import them then.

spaceone committed:
Add error flags for throwing errors back into generator

The following code demonstrates this behavior. As foo.on_error is
'throw_last' the on_error_1 handler is executed as well as on_error_2
which is the exception throwed back to the bar handler.

from circuits import Event, Component, handler, Debugger
import pytest

class foo(Event):
	on_error = 'throw_last'

class bar(Event):
	pass

class Foo(Component):
	@handler('foo', priority=2)
	def _on_error_1(self):
		raise ValueError('test')

	@handler('foo', priority=1)
	def _on_error_2(self):
		raise ValueError('test2')

	def bar(self):
		with pytest.raises(ValueError) as exc:
			x = yield self.call(foo())
		assert exc.value.args[0] == 'test2'
		print('DONE')

x = Foo()
x += Debugger()
x.fire(bar())
x.run()

spaceone committed:
specific handlers can overwrite error handling settings

spaceone commented:
@prologic Why do you close this and remove the branch?

prologic commented:
Ahh I was doing some Github cleanup! Sorry, do you want to take over the branch/PR?

spaceone commented:
I can't promise to work on this currently, but I still would need it. I restore the branch to not remove the current working state.
